### PR TITLE
S4158 FP: Collection filled in a for loop reported to be empty

### LINKED ISSUE #6179: S4158 FP: Collection filled in a for loop reported to be empty
### Description

Sonar Analyzer incorrectly reports S4158 for the code below. My original code is actually more complex. However, this simplified example also reports S4158.

### Repro steps

    public void FalsePositive()
    {
        var data = new[] { 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0 };
        var rising = new List<List<int>>();
        List<int> currentBin = null;
        for (var i = 0; i < data.Length; i++)
        {
            if (data[i] == 0)
            {
                if (currentBin != null)
                {
                    rising.Add(currentBin);
                    currentBin = null;
                }
        
                continue;
            }
        
            currentBin ??= new List<int>();
            currentBin.Add(i);
        }
        
        for (var i = 0; i < rising.Count; i++)
        {
            // SonarAnalyzer reports S4158 here:
            Console.WriteLine($"{rising[i].Count}");
        }
    }

### Expected behavior

As the collection is not empty, no issue should be reported.

### Actual behavior

S4158 is reported.

### Known workarounds

Using a foreach loop in the lower part, does not report the issue. However, in my original code this is not possible, as I actually have two collections that I have to access with indices.

### Related information

Environment:
    - Visual Studio Enterprise 17.3.0
    - SonarAnalyzer.CSharp 8.46.0.54807
    - .NET Framework 4.8
    - Windows 10 x64

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #6179

This has two parts.
For `var arr = new int[5]` we need to learn explicitly CollectionNotEmpty, so when the `for` condition of `i<arr.Length` arrives, we're sure we go inside and do not skip the loop

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs [hunk 1]
```diff
@@ -83,6 +84,11 @@ protected override ProgramState PreProcess(ProgramState state, IPropertyReferenc
             {
                 state = state.SetOperationValue(operation, value);
             }
+            if (operation.Property.Name is nameof(Array.Length) or nameof(List<int>.Count)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs:87]: While this is the correct thing to do, I have doubts if this file is the correct place to do it.  I'm considering doing this magic in the rule itself, because the constraint is rule-specific.
WDYT?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@
  */
 
 using SonarAnalyzer.SymbolicExecution.Constraints;
+using SonarAnalyzer.SymbolicExecution.Sonar.Constraints;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/References.cs:22]: This is wrong and needs a fix

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 3]
```diff
@@ -177,4 +175,23 @@ private ProgramState ProcessMethod(SymbolicContext context, IMethodSymbol method
         }
         return state;
     }
+
+    private static CollectionConstraint CollectionCreationConstraint(ProgramState state, IObjectCreationOperationWrapper objectCreation) =>
+        objectCreation.Arguments.SingleOrDefault(x => x.ToArgument().Parameter.Type.DerivesOrImplements(KnownType.System_Collections_IEnumerable)) is { } collectionArgument
+            ? state[collectionArgument]?.Constraint<CollectionConstraint>()
+            : CollectionConstraint.Empty;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:182]: This one is not changed, I just moved nonstatic ProcessMethod above the static ones and couldn't get the commit history play nicely about that

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 4]
```diff
@@ -184,4 +179,23 @@ private ProgramState ProcessMethod(SymbolicContext context, IMethodSymbol method
...
+
+    private static NumberConstraint PropertyReferenceConstraint(ProgramState state, IPropertyReferenceOperationWrapper propertyReference)
+    {
+        if (propertyReference.Property.Name is nameof(Array.Length) or nameof(List<int>.Count)
+            && propertyReference.Instance.TrackedSymbol() is { } symbol
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:191]: We could use `state.ResolveCapture(propertyReference.Instance).TrackedSymbol()` to cover more cases.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.CSharp8.cs [hunk 5]
```diff
@@ -153,3 +153,63 @@ public interface IDecoder
...
+            list.ForEach(x => { }); // Compliant
+        }
+    }
+
+    public void FilledInLoop_FromAnotherCollection_FromList()
+    {
+        var data = new List<int> { 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0 };
+        var list = new List<List<int>>();
+        List<int> currentBin = null;
+        for (var i = 0; i < data.Count; i++)
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.CSharp8.cs:193]: Should we also support `data.Count()` or do you want to do that in a different PR?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.CSharp8.cs [hunk 6]
```diff
@@ -153,3 +153,63 @@ public interface IDecoder
 {
     bool Convert(out int i3, out int i4);
 }
+
+// https://github.com/SonarSource/sonar-dotnet/issues/6179
+public class Repro_6179
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.CSharp8.cs:158]: Add some basic tests to demonstrate the logic without all the repro scaffolding. Also, add tests for LINQ-`.Count()`.

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.CSharp8.cs (hunk 5 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.CSharp8.cs:193]: I can do it now

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.vb [hunk 7]
```diff
@@ -193,7 +194,47 @@ Public Class CollectionTests
...
+        Else
+            Empty.Clear()  ' Compliant, unreachable
+        End If
+        NotEmpty.Clear()   ' Compliant, prevents LVA from throwing notEmpty away during reference capture
+    End Sub
+
+    Public Sub Enumerable_ExtensionIsInstance()
+        Dim Empty As New List(Of Integer)
+        Dim AsIEnumerable As IEnumerable(Of Integer) = Empty
+        If AsIEnumerable.Count = 0 Then ' This is invoked as instance, not as a static call
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.vb:233]: Educational: This a pure VB thing, right? Why is it calling this as instance?

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
The code needs to differentiate between the two `.Count()` overloads.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.vb (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.vb:233]: Yes, it's VB-only. 
There are some Roslyn internals about how they transform `IMethodSymbols` in a weird way sometimes. Not sure why exactly.

At least this one is understandable. For example [this thing](https://github.com/SonarSource/sonar-dotnet/blob/ba7978e771d0df3c83d841872821997b2702d392/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.vb#L179) is really weird.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 8]
```diff
@@ -148,13 +151,11 @@ public override void ExecutionCompleted()
...
-    private static CollectionConstraint CollectionCreationConstraint(ProgramState state, IObjectCreationOperationWrapper objectCreation) =>
-        objectCreation.Arguments.SingleOrDefault(x => x.ToArgument().Parameter.Type.DerivesOrImplements(KnownType.System_Collections_IEnumerable)) is { } collectionArgument
-            ? state[collectionArgument]?.Constraint<CollectionConstraint>()
-            : CollectionConstraint.Empty;
-
-    private static bool IsEmptyArray(IOperation operation) =>
-        operation.AsArrayCreation()?.DimensionSizes.Any(x => x.ConstantValue.Value is 0) ?? false;
+    private ProgramState ProcessInvocation(SymbolicContext context, IInvocationOperationWrapper invocation) =>
+        invocation.TargetMethod.Is(KnownType.System_Linq_Enumerable, nameof(Enumerable.Count))
+        && SizeConstraint(context.State, invocation.Instance ?? invocation.Arguments[0].ToArgument().Value) is { } constraint
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:156]: I'll split this, The `Count` thing is getting tricky as a new functionality

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs]: #7452

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM
